Add support to virtio-vDPA devices

amorenoz opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/ovn-org/ovn-kubernetes/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Trivial changes are exempt from following this template.
If your change is non-trivial, please provide the following information:
-->

**- What this PR does and why is it needed**
<!--
A summary of the changes within this pull request and some context
as to why they were made
-->

This PR works in combination with [this PR in the SR-IOV Network Device Plugin](https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin/pull/306) to provide support for vDPA devices.

This functionality is based on [OvS Hardware Offload](https://github.com/ovn-org/ovn-kubernetes/blob/master/docs/ovs_offload.md) and allows ovn-kubernetes to expose an open standard netdev such as a virtio-net to the pod.

Most of the heavy-lifting (detecting vdpa devices, ensuring they are bound the right driver, etc) is done by the SR-IOV Network Device Plugin. From a ovn-kubernetes perspective, it's just a matter of selecting the right netdev to move to the pod's namespace.


**- Special notes for reviewers**
<!--
What exactly did you change - you may also defer to information
contained in commit messages. At a bare minimum it's worth highlighting
which areas of the code were changed as it's easier to assign reviewers
-->
I would like feedback on a number of topics:
- Right now, this PR just detects if the PCI device has a virtio-vdpa device using [govdpa](https://github.com/redhat-virtio-net/govdpa) and exposes the virtio-net netdev if so. It is my understanding that supported cards, even if they allow the creation of two simultaneous netdevs (the vendor one and the virtio-vdpa device), the vdpa netdev takes precedence if it exists. Therefore, this logic would be sane.

If that is not the case we would need to add support to the [Device Information Spec](https://github.com/k8snetworkplumbingwg/device-info-spec/) so we can read the information coming from the Device Plugin and unambiguously determine the type of device we're handling
- I've placed the code that returns the virtio-net netdev in `pkg/util/sriovnet_linux.go`. It might not be the perfect place as it seems more of a wrapper around `https://github.com/Mellanox/sriovnet`. So this is a question for @moshe010 and @adrianchiris: WDYT? Would you accept this logic into `https://github.com/Mellanox/sriovnet`?


**- How to verify it**
<!--
Did you include unit tests? or end-to-end tests?
How can I manually verify that this patch achieves its objective
-->
You need a working OvS Hardware Offload setup with a ConnectX6-DX NIC. Then, follow the instructions that are added in `docs/ovs_offload.md` as part of this PR to configure virtio-vDPA devices.

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->
Add support for virtio-vdpa devices

amorenoz commented:
/cc @zshi-redhat

adrianchiris commented:
> ve placed the code that returns the virtio-net netdev in pkg/util/sriovnet_linux.go. It might not be the perfect place as it seems more of a wrapper around https://github.com/Mellanox/sriovnet. So this is a question for @moshe010 and @adrianchiris: WDYT? Would you accept this logic into https://github.com/Mellanox/sriovnet?

i think it makes sense. maybe govdpa should go to npwg.

amorenoz commented:
> > ve placed the code that returns the virtio-net netdev in pkg/util/sriovnet_linux.go. It might not be the perfect place as it seems more of a wrapper around https://github.com/Mellanox/sriovnet. So this is a question for @moshe010 and @adrianchiris: WDYT? Would you accept this logic into https://github.com/Mellanox/sriovnet?
> 
> i think it makes sense. maybe govdpa should go to npwg.

Agree, wasn't there an initiative to move some common stuff to a library in npwg that could be used by more projects?
I can propose govdpa to npwg in the next meeting

Should sriovnet also go to npwg too?
In any case I'll send a PR to add this part of vdpa code into sriovnet

amorenoz commented:
Hi @adrianchiris , as we discussed, having sriovnet depend on govdpa does not make much sense since the goal of vdpa is to (eventually) work with all possible vdpa parents (sriov is just one of them).

I've moved the govdpa-related stuff to an independent vdpa_linux.go wrapper so it can be mocked and added unit tests. PTAL

amorenoz commented:
Rebased on top of @girishmg 's latest change: https://github.com/ovn-org/ovn-kubernetes/commit/60eeb3ae424ef7ebb817233abc03b840983836a1

amorenoz commented:
@girishmg @dcbw @danwinship any comments on this?

dcbw commented:
@amorenoz could you rebase the PR? Looks OK to me, just needs the rebase.

lmilleri commented:
@dcbw I rebased and pushed the PR again, but got an error in the ovn-ci / e2e-dual-conversion. Hope this is an expected failure and not due to my changes?

zshi-redhat commented:
/retest

zshi-redhat commented:
> Oops, something went wrong:
> 
> ```
> Unable to retry this workflow run because it was created over a month ago
> ```

Interesting, retest doesn't work for long living PR, @dcbw do you know how to rerun the test w/o creating a new PR (the CI failure doesn't seem to be related to the vdpa change)?

zshi-redhat commented on go-controller/go.sum:
```diff
@@ -497,6 +497,8 @@ github.com/juju/utils v0.0.0-20180808125547-9dfc6dbfb02b/go.mod h1:6/KLg8Wz/y2KV
 github.com/juju/version v0.0.0-20161031051906-1f41e27e54f2/go.mod h1:kE8gK5X0CImdr7qpSKl3xB2PmpySSmfj7zVbkZFs81U=
 github.com/julienschmidt/httprouter v1.2.0/go.mod h1:SYymIcj16QtmaHHD7aYtjjsJG7VTCxuUUipMqKk8s4w=
 github.com/julienschmidt/httprouter v1.3.0/go.mod h1:JR6WtHb+2LUe8TCKY3cZOxFyyO8IZAc4RVcycCCAKdM=
+github.com/k8snetworkplumbingwg/govdpa v0.1.3 h1:FZRhTMB1e3yWwSEy+l4eS73WioyMaL+vmFZ8JNwn+Uk=
```
@lmilleri I vaguely remembered that you mentioned we need latest changes in govdpa lib, but this is still using 0.1.3 version, is this expected?

trozet commented:
> > Oops, something went wrong:
> > ```
> > Unable to retry this workflow run because it was created over a month ago
> > ```
> 
> Interesting, retest doesn't work for long living PR, @dcbw do you know how to rerun the test w/o creating a new PR (the CI failure doesn't seem to be related to the vdpa change)?

@zshi-redhat I can restart it, but i think it wont restart now because there are conflicting files. If this PR is still important can you rebase @amorenoz and we will review.

zshi-redhat commented:
@lmilleri could you please rebase?

lmilleri commented:
@zshi-redhat there are some changes in the govdpa library under review: https://github.com/k8snetworkplumbingwg/govdpa/pull/15.
I think it would make sense to push the above PR first, create a new version and then rebase this PR afterwards

lmilleri commented:
Rebased from master and updated dependency to the newly govdpa v0.1.4 release.
Please review

lmilleri commented:
Please consider this PR on hold, I found a potential issue in govdpa v0.1.4

lmilleri commented:
Fixed integration issue with govdpa v.0.1.4.
Please review.

@zshi-redhat  @dcbw

zshi-redhat commented:
/retest

lmilleri commented:
@trozet @dcbw do you think this PR can be merged?
If not, please advise on the way forward.
Apparently "/retest" is not working, any workaround?

amorenoz commented:
@trozet @zshi-redhat @dcbw Can you please take a look at this PR?

dcbw commented:
/lgtm

mostly just helper code

lmilleri commented:
Yes, it is mostly helper code apart from go-controller/pkg/cni/cni.go, in which resides the logic for returning the virtio/vdpa device if vdpa device exists.

lmilleri commented:
@girishmg can you please review it?
